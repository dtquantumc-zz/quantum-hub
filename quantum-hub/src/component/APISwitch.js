// SPDX-License-Identifier: MIT
/*
 * (C) Copyright 2020
 * Diversifying Talent in Quantum Computing, Geering Up, UBC
 */

import React from 'react'
import PropTypes from 'prop-types'
import './component_css/APISwitch.css'
import ToggleSwitch from './toggleSwitch'

/**
 * APISwitch is a simple React Component
 * Like the toggleSwitch, but it has a textbox that's
 * active only if the switch itself is active.
 * Passing it a function for setting API key lets it
 * update a parent's API key (to '' if disabled or empty,
 * otherwise to the exact inputted string).
 *
 * To modify the look of the textbox, modify APISwitch.css
 * For the Switch, please look at ToggleSwitch
 *
 * Example Use::
 *
 *  <APISwitch
 *    id='myAPISwitch'
 *    switchText={['QPU2', 'Sim2']}
 *    setAPIKey={(key) => {
 *      console.log(key)
 *    }}
 *    defaultChecked={false}
 *  />
 *
 * Properties:
 * @prop {String} id - (Required) All Document ids should be different, this is no different.
 * @prop {Array(String)} switchText - (Required) switchText controls the text on the actual switch.
 * @prop {Func(String)} setAPIKey - (Required) Any time the API key is updated, this is called.
 * Use it to set parent values, so parent has access to this textbox's values :D
 * @prop {Bool} defaultChecked - Set if you want the switch to be on by default.
 */
class APISwitch extends React.Component {
  static propTypes = {
    /** The inner switch and textbox have ids <id>+'_switch' and <id>+'_text' respectively. */
    id: PropTypes.string.isRequired,
    /** switchText is required. Eg. ['On', 'Off'] */
    switchText: PropTypes.arrayOf(PropTypes.string).isRequired,
    setAPIKey: PropTypes.func.isRequired,
    default: PropTypes.bool
  }

  constructor (props) {
    super(props)
    this.state = {
      innerText: '',
      disabled: !props.defaultChecked
    }
  }

  render () {
    return (
      <div className='api_switch'>
        {this.props.id ? (
          <span id={this.props.id}>
            <ToggleSwitch
              id={this.props.id + '_switch'}
              Text={this.props.switchText}
              Name='APISwitch'
              onChange={(checked) => {
                this.setState({
                  innerText: this.state.innerText,
                  disabled: !checked
                })
                this.props.setAPIKey(
                  checked ? this.state.innerText : ''
                )
              }}
              defaultChecked={this.props.defaultChecked}
              disabled={false}
            />
            <input
              id={this.props.id + '_text'}
              name='APIText'
              type='text'
              className='api_text_box'
              disabled={this.state.disabled}
              onChange={(e) => {
                this.setState({
                  innerText: e.target.value,
                  disabled: this.state.disabled
                })
                this.props.setAPIKey(e.target.value)
              }}
            />
          </span>
        ) : null}
      </div>
    )
  }
}

export default APISwitch
