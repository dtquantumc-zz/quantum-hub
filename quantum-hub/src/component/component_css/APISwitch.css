.api_text_box{
  width: 150px;
}

.api_text_box:disabled{
  cursor: not-allowed;
  background-color: #EEEEEE;
}

.api_text_box:focus {
  outline: #3399FF auto 5px;
}