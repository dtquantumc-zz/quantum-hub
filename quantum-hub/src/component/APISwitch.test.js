import React from 'react'
import APISwitch from './APISwitch'
// import renderer from 'react-test-renderer'
// import { cleanup, fireEvent } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import { render, unmountComponentAtNode } from 'react-dom'
import userEvent from '@testing-library/user-event'

let container = null
beforeEach(() => {
  // Setup DOM element as render target
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  // cleanup
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

it('looks the same as last time (snapshot)', () => {
  const setAPI = jest.fn((key) => { return key })
  act(() => {
    render(
      <APISwitch
        defaultChecked={false}
        id='mySwitch'
        switchText={['on', 'off']}
        setAPIKey={setAPI}
      />,
      container)
  })
  // Get whole thing, make sure it hasn't fundamentally changed
  const apiS = document.getElementById('mySwitch')
  expect(apiS).toMatchSnapshot()
})

it('can be typed in and disabled', async () => {
  const setAPI = jest.fn((key) => { return key })
  act(() => {
    render(
      <APISwitch
        defaultChecked
        id='mySwitch'
        switchText={['on', 'off']}
        setAPIKey={setAPI}
      />,
      container)
  })

  // get input element within APISwitch
  const inpBox = document.getElementById('mySwitch_text')
  // const apiS = document.getElementById('mySwitch')
  const togSwitch = document.getElementById('mySwitch_switch')
  expect(inpBox.value).toBe('')

  // trigger typing
  // Note, it types one character at a time :O
  const testStr = 'Key:234'
  await userEvent.type(inpBox, testStr)

  // Check the right num of calls
  expect(setAPI).toHaveBeenCalledTimes(testStr.length)

  // Check each call input and output is correct
  for (let i = 0; i < testStr.length; ++i) {
    const testSlice = testStr.slice(0, i + 1)
    expect(setAPI.mock.calls[i][0]).toBe(testSlice)
    expect(setAPI.mock.results[i].value).toBe(testSlice)
  }

  // Click the switch to disable then enable the textbox
  userEvent.click(togSwitch)
  userEvent.click(togSwitch)

  // Check that the number of calls has increased by two
  expect(setAPI).toHaveBeenCalledTimes(testStr.length + 2)

  // Check correctness of calls
  expect(setAPI.mock.calls[testStr.length][0]).toBe('')
  expect(setAPI.mock.results[testStr.length].value).toBe('')

  expect(setAPI.mock.calls[testStr.length + 1][0]).toBe(testStr)
  expect(setAPI.mock.results[testStr.length + 1].value).toBe(testStr)
})

it('doesn\'t get typed in when switched off', async () => {
  const setAPI = jest.fn((key) => { return key })
  act(() => {
    render(
      <APISwitch
        id='mySwitch'
        switchText={['on', 'off']}
        setAPIKey={setAPI}
      />,
      container)
  })

  const inpBox = document.getElementById('mySwitch_text')

  // get ToggleSwitch element (actually gets internal checkbox)
  const toggleS = document.getElementById('mySwitch_switch')
  expect(toggleS.checked).toBe(false)

  // trigger type, but now its disabled!
  const testStr = 'Key:234'
  await userEvent.type(inpBox, testStr)

  // Check the right num of calls
  expect(setAPI).toHaveBeenCalledTimes(0)

  // Check that there is no text typed
  expect(inpBox.value).toBe('')
})
