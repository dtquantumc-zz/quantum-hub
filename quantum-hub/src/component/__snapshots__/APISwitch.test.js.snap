// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`looks the same as last time (snapshot) 1`] = `
<span
  id="mySwitch"
>
  <div
    class="toggle-switch"
  >
    <input
      class="toggle-switch-checkbox"
      id="mySwitch_switch"
      name="APISwitch"
      type="checkbox"
    />
    <label
      class="toggle-switch-label"
      for="mySwitch_switch"
    >
      <span
        class="toggle-switch-inner"
        data-no="off"
        data-yes="on"
      />
      <span
        class="toggle-switch-switch"
      />
    </label>
  </div>
  <input
    class="api_text_box"
    disabled=""
    id="mySwitch_text"
    name="APIText"
    type="text"
  />
</span>
`;
